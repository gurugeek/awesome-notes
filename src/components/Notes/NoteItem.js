import React, { Component } from 'react';
import { View, Text, Animated } from 'react-native';
import PropTypes from 'prop-types';
import Swipeout from 'react-native-swipeout';

import { NotesStyle } from '../../styles';
import NoteButton from './NoteButton';

class NoteItem extends Component {
    constructor(props) {
        super(props);
        this.state = {
            scaleAnim: new Animated.Value(1),
            opacityAnim: new Animated.Value(1)
        };

        this._switchFavorite = this._switchFavorite.bind(this);
        this._onSwipeOpen = this._onSwipeOpen.bind(this);
        this._onSwipeClose = this._onSwipeClose.bind(this);
    }

    _runScaleAnimation(toValue) {
        Animated.timing(this.state.scaleAnim, { toValue }).start();
    }

    _runOpacityAnimation(toValue) {
        Animated.timing(this.state.opacityAnim, { toValue }).start();
    }

    _formatDate(date) {
        const [day, month, year] = ['' + date.getDate(), '' + (date.getMonth() + 1), date.getFullYear()];
        return [day, month, year].join('/');
    }

    _switchFavorite() {
        this.props.note.favourite = !this.props.note.favourite;
        this.props.updateNote(this.props.note);
    }

    _onSwipeOpen(sectionID, rowId, direction) {
        this._runScaleAnimation(0.60);
        direction == 'right' && this._runOpacityAnimation(0.5);
    }

    _onSwipeClose(sectionID, rowId, direction) {
        this._runScaleAnimation(1);
        direction === 'right' && this._runOpacityAnimation(1);
    }

    render() {
        const swipeSettings = {
            autoClose: true,
            right: [{ onPress: () => { this.props.deleteNote(this.props.note.id)}, component: <NoteButton type='delete' />, backgroundColor: 'transparent' }],
            left: [{ onPress: this._switchFavorite, component: <NoteButton type={this.props.note.favourite ? 'favourite-selected' : 'favourite'} />, backgroundColor: 'transparent'}],
            rowId: this.props.note.id,
            onOpen: this._onSwipeOpen,
            onClose: this._onSwipeClose,
            style: NotesStyle.noteItem.swipeoutWrapper
        };

        let { scaleAnim, opacityAnim } = this.state;
        const rotateY = this.state.opacityAnim.interpolate({
            inputRange: [0.5, 1],
            outputRange: ['45deg', '0deg']
        });
        return (
            <Swipeout {...swipeSettings}>
                <Animated.View style={[NotesStyle.noteItem.wrapper, { opacity: opacityAnim }, { transform: [{ scale: scaleAnim }, { perspective: 1000 }, { rotateY }] }]}>
                    <View style={NotesStyle.noteItem.textWrapper}>
                        <Text style={NotesStyle.noteItem.text}>{this.props.note.text}</Text>
                    </View>
                    <View style={NotesStyle.noteItem.dateWrapper}>
                        <Text style={NotesStyle.noteItem.date}>{this._formatDate(this.props.note.createdAt)}</Text>
                    </View>
                </Animated.View>
            </Swipeout>
        );
    }
}

NoteItem.propTypes = {
    note: PropTypes.object.isRequired,
    updateNote: PropTypes.func.isRequired,
    deleteNote: PropTypes.func.isRequired
};

export default NoteItem;