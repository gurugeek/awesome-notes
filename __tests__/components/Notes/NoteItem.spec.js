import React from 'react';
import { View, Text, Animated } from 'react-native';
import Swipeout from 'react-native-swipeout';
import { shallow } from 'enzyme';
import renderer from 'react-test-renderer';
import Enzyme from 'enzyme';
import Adapter from 'enzyme-adapter-react-16';

Enzyme.configure({ adapter: new Adapter() });

import NoteItem from '../../../src/components/Notes/NoteItem';

jest.useFakeTimers();

const note = {
   id: 'awesomeId',
   text: 'Awesome text',
   favourite: false,
   createdAt: new Date(601254000000) // my birthday ^^
};

describe('<NoteItem /> Component', () => {
    it('NoteItem should match with snapshot', () => {
        const tree = renderer.create(
            <NoteItem note={note} updateNote={() => {}} deleteNote={() => {}} />
        );
        const json = tree.toJSON();
        expect(json).toMatchSnapshot();
    });

    const noteItemShallow = shallow(<NoteItem note={note} updateNote={() => {}} deleteNote={() => {}} />);

    it('NoteItem should exists', () => {
        expect(noteItemShallow.length).toEqual(1);
    });

    it('NoteItem should has two View', () => {
        expect(noteItemShallow.find(View)).toHaveLength(2);
    });

    it('NoteItem should has one Swipeout', () => {
        expect(noteItemShallow.find(Swipeout)).toHaveLength(1);
    });

    it('NoteItem should has one Animated.View', () => {
        expect(noteItemShallow.find(Animated.View)).toHaveLength(1);
    });

    it('NoteItem should has two Text', () => {
        expect(noteItemShallow.find(Text)).toHaveLength(2);
    });
});