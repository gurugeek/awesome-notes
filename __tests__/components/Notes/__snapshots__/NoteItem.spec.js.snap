// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<NoteItem /> Component NoteItem should match with snapshot 1`] = `
<View
  style={
    Array [
      Object {
        "backgroundColor": "#dbddde",
        "overflow": "hidden",
      },
      Object {
        "backgroundColor": "transparent",
        "marginBottom": 10,
      },
    ]
  }
>
  <View
    onLayout={[Function]}
    onMoveShouldSetResponder={[Function]}
    onMoveShouldSetResponderCapture={[Function]}
    onResponderEnd={[Function]}
    onResponderGrant={[Function]}
    onResponderMove={[Function]}
    onResponderReject={[Function]}
    onResponderRelease={[Function]}
    onResponderStart={[Function]}
    onResponderTerminate={[Function]}
    onResponderTerminationRequest={[Function]}
    onStartShouldSetResponder={[Function]}
    onStartShouldSetResponderCapture={[Function]}
    style={
      Array [
        Object {},
        Object {
          "left": 0,
        },
      ]
    }
  >
    <View
      collapsable={undefined}
      style={
        Object {
          "backgroundColor": "#A8DADC",
          "borderRadius": 10,
          "opacity": 1,
          "padding": 10,
          "transform": Array [
            Object {
              "scale": 1,
            },
            Object {
              "perspective": 1000,
            },
            Object {
              "rotateY": "0deg",
            },
          ],
        }
      }
    >
      <View
        style={
          Object {
            "flex": 1,
            "marginBottom": 5,
          }
        }
      >
        <Text
          accessible={true}
          allowFontScaling={true}
          ellipsizeMode="tail"
          style={
            Object {
              "color": "#211A1E",
              "fontFamily": "Roboto-Regular",
            }
          }
        >
          Awesome text
        </Text>
      </View>
      <View
        style={
          Object {
            "height": 10,
          }
        }
      >
        <Text
          accessible={true}
          allowFontScaling={true}
          ellipsizeMode="tail"
          style={
            Object {
              "color": "#457B9D",
              "fontFamily": "Roboto-Thin",
              "fontSize": 10,
              "textAlign": "right",
            }
          }
        >
          20/1/1989
        </Text>
      </View>
    </View>
  </View>
  <View />
  <View />
</View>
`;
